### ISSUE TITLE
Consider adding a way to customize console log output format

### ISSUE THREAD

**Issue by robbieknuth** [aspnet/Extensions#2276]
### Is your feature request related to a problem? Please describe.
The default logging format is good for readability. And it is already demonstrated that customization is needed, given the existence of ConsoleLoggerFormat, eg Default and Systemd. I sometimes have a need for logging raw text without any kind of formatting at all, and sometimes to output a json formatted log for more structured logging. I want a way to take the components of the log messages and format them in whatever way i see fit, but benefit from the work done to have a log queue that prevents log tearing, and be able to plug this logger into everything that expects an ILogger already.

### Describe the solution you'd like
I would like a way to specify the format used by the console logger, to potentially eliminate timestamps, event id, etc, and simply write text. Or, take a log message with its various fields and return a string

### Describe alternatives you've considered
- Console.Writeline => this works, but doesn't play nice with DI, nor does it play nice with things like LogLevel. Also the console logger has a queue to prevent log tearing.
- Heavily copying the consolelogger implementation => this isn't that hard, and will be what I do in the short term, however something built in would eliminate the need for a custom implementation. And special casing Systemd acknowledges that custom formatting is already needed.

### Additional context
Effectively I want a way to give a delegate that would act like CreateSystemdLogMessage or CreateDefaultLogMessage.

**Comment by nvoskuilen** [comment]
Consider taking a look at [JsonConsole](https://github.com/nvoskuilen/Nvosk.Extensions.Logging.JsonConsole) which solves the "json formatted log" question.

Based on Microsoft.Extensions.Logging.Console v3.0.0 with added json formatted log output utilizing the new System.Text.Json.Utf8JsonWriter.

Sample output:
```javascript
{"time":"2019-12-31T23:59:54","level":"trce","source":"ConsoleApp.Program","event_id":0,"message":"LogTrace..."}
{"time":"2019-12-31T23:59:55","level":"dbug","source":"ConsoleApp.Program","event_id":0,"message":"LogDebug..."}
{"time":"2019-12-31T23:59:56","level":"info","source":"ConsoleApp.Program","event_id":0,"message":"LogInformation..."}
{"time":"2019-12-31T23:59:57","level":"warn","source":"ConsoleApp.Program","event_id":0,"message":"LogWarning..."}
{"time":"2019-12-31T23:59:58","level":"fail","source":"ConsoleApp.Program","event_id":0,"message":"LogError..."}
{"time":"2019-12-31T23:59:59","level":"crit","source":"ConsoleApp.Program","event_id":0,"message":"LogCritical..."}
```

DotNet dev team might consider to merge JsonConsole into the next release.

**Comment by pharring** [comment]
The JSON formatting part of this suggestion is covered by #2479

**Comment by anurse** [comment]
I think this will become much more relevant as the way to implement things like #2479 . Now that we have 2 formats (Systemd and Default) and are looking to add two more (Singleline and Json), a "formatter" abstraction starts to make a lot more sense.

**Comment by pharring** [comment]
Yes, what I discovered from writing my own structured console logger was that the bulk of the work was in the serializer (formatter?). That the serialized bytes ended up going to stdout was practically insignificant.

My `JsonLogger` implements `ILogger` and emits spans of (serialized UTF8) bytes to a 'sink':
```csharp
    /// <summary>
    /// An interface over a sink for bytes.
    /// </summary>
    public interface ISpanWriter
    {
        /// <summary>
        /// Write the given span of bytes to the sink.
        /// </summary>
        /// <param name="span">The bytes to write.</param>
        void Write(ReadOnlySpan<byte> span);
    }
```
There's an abstract `JsonLoggerProvider` that knows everything except the sink (`ISpanWriter`) and a concrete `StructuredConsoleLoggerProvider` derived from it to supply a sink over stdout.
All the "options" regarding JSON structure (timestamp format; which properties to include/ignore; line separator; JSON converter options) are in the `JsonLoggerProvider`.

**Comment by anurse** [comment]
We expect to do this as part of the work on structured logging in dotnet/runtime#34742
